Compression/src/CompressDetStepMCs_module.cc

```diff
@@ -0,0 +1,638 @@
+      recordSimParticle(mcTrajectory.second.sim());
+    }
+    else {
+      throw cet::exception("CompressDetStepMCs") << "Unrecognized compression level \"" << _mcTrajectoryCompressionLevel.name() <<"\" for MCTrajectories" << std::endl;
+    }
+  }
+}
+
+void mu2e::CompressDetStepMCs::updateStepPointMCs() {
+  for (const auto& i_tag : _conf.stepPointMCTags()) {
+    for (auto& i_stepPointMC : *(_newStepPointMCs.at(i_tag.instance()))) {
+      const auto& oldSimPtr = i_stepPointMC.simParticle();
+      art::Ptr<mu2e::SimParticle> newSimPtr = _simPtrRemap.at(oldSimPtr);
+      if(_conf.debugLevel()>0) {
+        std::cout << "Updating SimParticlePtr in StepPointMC from " << oldSimPtr << " to " << newSimPtr << std::endl;
+      }
+      i_stepPointMC.simParticle() = newSimPtr;
+    }
+  }
+}
+
+void mu2e::CompressDetStepMCs::updateMCTrajectories() {
+  for (auto& i_mcTrajectory : *_newMCTrajectories) {
+    const auto& oldSimPtr = i_mcTrajectory.second.sim();
+    art::Ptr<mu2e::SimParticle> newSimPtr = _simPtrRemap.at(oldSimPtr);
+    if(_conf.debugLevel()>0) {
+      std::cout << "Updating SimParticlePtr in MCTrajectory from " << oldSimPtr << " to " << newSimPtr << std::endl;
+    }
+    i_mcTrajectory.second.sim() = newSimPtr;
+    auto mcTrajPair = _newMCTrajectories->extract(i_mcTrajectory.first);
+    mcTrajPair.key() = newSimPtr;
+    _newMCTrajectories->insert(std::move(mcTrajPair));
+    //    i_mcTrajectory.first = newSimPtr;
+  }
+}
+
+void mu2e::CompressDetStepMCs::recordSimParticle(const art::Ptr<mu2e::SimParticle>& sim_ptr) {
+  // Also need to add all the parents too
+  _simParticlesToKeep[sim_ptr.id()].insert(sim_ptr);
+  art::Ptr<mu2e::SimParticle> childPtr = sim_ptr;
+  art::Ptr<mu2e::SimParticle> parentPtr = childPtr->parent();
+
+  if(_conf.debugLevel()>0) {
+    std::cout << "Recording SimParticle " << sim_ptr << std::endl;
+  }
+  while (parentPtr) {
+    MCRelationship mcr(sim_ptr, parentPtr);
+    if (_keepNGenerations == -1 || ( (mcr.removal() <= _keepNGenerations) && mcr.removal()>=0) ) {
+      _simParticlesToKeep[sim_ptr.id()].insert(parentPtr);
+      if(_conf.debugLevel()>0) {
+        std::cout << "and recording its ancestor " << parentPtr << " (NGen = " << (int)mcr.removal() << ")" << std::endl;
+      }
+    }
+    else if (parentPtr->isPrimary()) { // always keep the very first SimParticle
+      _simParticlesToKeep[sim_ptr.id()].insert(parentPtr);
+      if(_conf.debugLevel()>0) {
+        std::cout << "and recording the very first SimParticle " << parentPtr << std::endl;
+      }
+    }
+    else {
+      if(_conf.debugLevel()>0) {
+        std::cout << "and *not* recording its ancestor " << parentPtr << " (NGen = " << (int)mcr.removal() << ")" << std::endl;
+      }
+    }
+    childPtr = parentPtr;
+    parentPtr = parentPtr->parent();
+  }
+}
+
+void mu2e::CompressDetStepMCs::checkCompressionLevels() {
```

Comment:
I am confused why step compression level is an option when there is only 1 allowed value.  Do you have a use case in mind where we would actually remove steps?  I can't think of one.

Reply:
I don't see us removing steps in this module. I like to be explicit about what we're touching and what we're doing and, if we end up refactoring CompressDigiMCs to match this module more, it will provide some consistency.